### PR TITLE
[Calyx] Parse attributes for BlockArguments and emit Calyx attributes.

### PR DESCRIPTION
**Pull request by cgyurgyik**
This PR does not address attributes for `Cell`s. The work necessary for this is discussed in #1744.

### CODE REVIEW

**Comment by cgyurgyik** [comment]
A bit of discussion:

While reading up on attributes, I learned that only arguments and results of `FunctionLike` operations may have attribute dictionaries. As of right now, this means that we can't attach attribute dictionaries directly to the results of a primitive operation, such as `RegisterOp`.

Next, I'd like some feedback on representing Calyx's version of attributes. A Calyx attribute is, an identifier followed by an integer. E.g. `@bound(5)` or `"static"=1`.

In CIRCT, each argument/result of a `FunctionLike` operation may optionally have a dictionary of {Identifier, Attribute}. E.g.
```
// Below, the Identifier `go` maps to Unit. 
// This could map to, say, an IntegerAttr, similar to Calyx attributes. This would look like { clk = 1 }

calyx.component @main(%go: i1 {go}, ... ) -> ( ... ) { }
```

To avoid some level of brittle-ness, we could also verify that only Calyx-representable attributes are attached to certain operations and/or the arguments/results of a component. For example, it would be non-sensical to attach `@bound(5)` to a GroupOp.


Curious to hear your thoughts. @mikeurbach @mortbopet @rachitnigam

**Comment by rachitnigam** [comment]
> In CIRCT, each argument/result of a FunctionLike operation may optionally have a dictionary of {Identifier, Attribute}.

I think that, for now, limiting the representation to allow for `IntegerAttr` sounds good. If, in the future, the native compiler allows for more forms of attributes, we can enable other kinds of attributes.

> To avoid some level of brittle-ness, we could also verify that only Calyx-representable attributes are attached to certain operations and/or the arguments/results of a component.

I'm not so sure about this. Attributes, but their nature, do not have any semantics--the passes choose to interpret them in whatever way they want. Right now, there is no verifier in the Calyx compiler to check if something like `@bound` is attached to a port.

Is it the case that CIRCT requires this? That is, does CIRCT require us to specify the exact set of attributes we want to use a priori?

**Comment by cgyurgyik** [comment]
It is not the case. It would be a check that a developer doesn't misunderstand the designated purpose of a certain attribute. Might be overkill though.

**Comment by cgyurgyik** [comment]
Added support for emitting component port attributes. Still need to support attributes for the components themselves, groups, and control constructs.

**Comment by mikeurbach** [comment]
FWIW the way `FunctionLike` operations have result attributes is pretty simple: under the hood it's just a DictionaryAttr with keys like `result0`, `result1`, etc. (actually, it may have switched to an ArrayAttr with integer keys). It's nothing magic, and we could use a similar approach for attaching attributes to the results of primitives.

The work in progress looks good to me, by the way.

**Comment by cgyurgyik** [comment]
This is mostly done. The last thing I need to do is make changes to the `CellInterface` to support attributes with `Cell`s. For sake of PR size, I think that can be done separately.

**Review by mortbopet — APPROVED** [review]
Looks good to me - thank you for also doing a bit of cleanup along the way :+1:. I think it would be good to leave the cell interface refactoring to another commit.

File: include/circt/Dialect/Calyx/CalyxInterfaces.td [hunk 1]
```diff
@@ -91,9 +91,15 @@ def CellOpInterface : OpInterface<"CellInterface"> {
       /*methodBody=*/"",
       /*defaultImplementation=*/[{
         SmallVector<circt::calyx::PortInfo> info;
-        for (auto it : llvm::zip($_op->getResults(), $_op.portDirections(),$_op.portNames()))
-          info.push_back(PortInfo{StringAttr::get($_op->getContext(),
-            std::get<2>(it)), std::get<0>(it).getType(), std::get<1>(it)});
+
+        auto zip = llvm::zip($_op->getResults(), $_op.portDirections(), $_op.portNames());
+        for (auto&& [result, direction, name] : zip)
```
Comment (mikeurbach) [inline include/circt/Dialect/Calyx/CalyxInterfaces.td:96]: Is this a C++17 feature? I hadn't seen it before but it's nice.

File: include/circt/Dialect/Calyx/CalyxInterfaces.td [hunk 2]
```diff
@@ -91,9 +91,15 @@ def CellOpInterface : OpInterface<"CellInterface"> {
...
+          info.push_back(PortInfo{
+              .name=StringAttr::get($_op->getContext(), name),
```
Comment (mikeurbach) [inline include/circt/Dialect/Calyx/CalyxInterfaces.td:98]: I think naming the struct members in the initializer like this is a C++20 feature (and we're on 17 for now), do you get a warning about that?

**Review by mikeurbach — APPROVED** [review]
Looks good to me, just a couple questions about some of the slick C++ syntax.

File: include/circt/Dialect/Calyx/CalyxInterfaces.td (hunk 1 above)
Comment (cgyurgyik) [reply include/circt/Dialect/Calyx/CalyxInterfaces.td:96]: Yep! https://en.cppreference.com/w/cpp/language/structured_binding

File: include/circt/Dialect/Calyx/CalyxInterfaces.td (hunk 2 above)
Comment (cgyurgyik) [reply include/circt/Dialect/Calyx/CalyxInterfaces.td:98]: No warnings AFAICT. I've been doing this for a at least a year or two now. Here's the [cppreference](https://en.cppreference.com/w/c/language/struct_initialization) on struct initialization, if I'm understanding correctly what you're referencing:

Comment (mikeurbach) [reply include/circt/Dialect/Calyx/CalyxInterfaces.td:98]: Ok great. I was thinking of "designated initializers" (syntax (3) and (4) here: https://en.cppreference.com/w/cpp/language/aggregate_initialization).

**Pull request merged**